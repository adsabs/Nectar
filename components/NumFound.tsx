import { Typography, Box } from '@material-ui/core';
import { useRecoilValue } from 'recoil';
import { numFoundState } from '@recoil/selectors';

const NumFound: React.FC<INumFoundProps> = () => {
  const numFound = useRecoilValue(numFoundState);

  return (
    <Typography variant="caption" component="article">
      Your search returned{' '}
      <Box fontWeight="fontWeightMedium" display="inline">
        {numFound.toLocaleString()}
      </Box>{' '}
      results
    </Typography>
  );
};

interface INumFoundProps {}

export default NumFound;
