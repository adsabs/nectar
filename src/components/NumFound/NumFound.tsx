import { Box, SkeletonText, Text } from '@chakra-ui/react';
import { useStore } from '@/store';
import { ReactElement } from 'react';

import { truncateDecimal } from '@/utils/common/formatters';
import { useGetSearchStats } from '@/api/search/search';

export interface INumFoundProps {
  count?: number;
  isLoading?: boolean;
}

// pin lower bound to 0, so we don't get negative numbers, and convert to locale string
const sanitizeNum = (num: number): string => {
  return (num < 0 ? 0 : num).toLocaleString();
};

export const NumFound = (props: INumFoundProps): ReactElement => {
  const { count = 0, isLoading } = props;

  if (isLoading) {
    return (
      <Box h={5}>
        <SkeletonText noOfLines={1} w="40" mt="1" skeletonHeight={2} />
      </Box>
    );
  }

  const countString = typeof count === 'number' ? sanitizeNum(count) : '0';

  return (
    <Box h={5}>
      <Text role="status" fontSize="xs">
        Your search returned{' '}
        <Text as="span" fontWeight="bold">
          {countString}
        </Text>{' '}
        results <SortStats />
      </Text>
    </Box>
  );
};

const SortStats = () => {
  const latestQuery = useStore((state) => state.latestQuery);
  const { data, isSuccess } = useGetSearchStats(latestQuery);

  if (isSuccess && 'citation_count' in data.stats_fields) {
    const count = sanitizeNum(data.stats_fields.citation_count.sum);
    return (
      <>
        with{' '}
        <Text as="span" fontWeight="bold" fontSize="xs">
          {count}
        </Text>{' '}
        total citations
      </>
    );
  }

  if (isSuccess && 'citation_count_norm' in data.stats_fields) {
    const count = sanitizeNum(truncateDecimal(data.stats_fields.citation_count_norm.sum, 2));
    return (
      <>
        with{' '}
        <Text as="span" fontWeight="bold" fontSize="xs">
          {count}
        </Text>{' '}
        total normalized citations
      </>
    );
  }
  return null;
};
