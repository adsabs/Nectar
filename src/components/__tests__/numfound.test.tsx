import { render } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import React from 'react';
import NumFound from '../numfound';
expect.extend(toHaveNoViolations);

test('renders correctly with no props', async () => {
  const { container } = render(<NumFound />);
  expect(container).toMatchSnapshot();
  expect(container).toHaveTextContent('Your search returned 0 results');
});

test('renders correctly with prop', async () => {
  const { container } = render(<NumFound numFound={10} />);
  expect(container).toHaveTextContent('Your search returned 10 results');
});

test('Has no accessibility violations', async () => {
  render(<NumFound />);
  const results = await axe(document.body);
  expect(results).toHaveNoViolations();
});
