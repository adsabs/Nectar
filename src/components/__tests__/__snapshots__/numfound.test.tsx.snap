// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with no props 1`] = `
<div>
  <p
    class="text-xs"
    role="status"
  >
    Your search returned
     
    <span
      class="font-bold"
    >
      0
    </span>
     results
  </p>
</div>
`;
